### PR TITLE
tracee-ebpf: flatten event probe dependencies

### PR DESCRIPTION
**Pull request by yanivagman**
## Initial Checklist

- [ ] There is an issue describing the need for this PR.
- [x] Git log contains summary of the change.
- [x] Git log contains motivation and context of the change.
- [ ] If part of an EPIC, PR git log contains EPIC number.
- [ ] If part of an EPIC, PR was added to EPIC description.

## Description (git log)

Currently, in order to define which probes are required for an event to
function correctly, we either define the probes explicitly or use an
event dependency, which will load and attach the required probes.
This other method of defining the probes implicitly should be used only
when the event is a real dependency, and not just as a mechanism to load
and attach the other event's bpf program.

This commit explicitly defines which probes are required for an event to
function correctly, and always submit an event to userspace when it is
defined as a dependency of another event. In addition, it sets tracee to
always load all the bpf programs that exists on the BPF object file.
This is required to break the coupling between an event and the probes
used by it. It will also be required to support dynamically changing
events during runtime as described in [1]. Another future PR might not
load all the BPF probes if it will detect that not all the required BPF
dependencies are met (e.g. LSM-BPF is not supported by the running
kernel).

## Type of change

- [x] New feature (non-breaking change adding functionality).

## How Has This Been Tested?

Manually

Reproduce the test by running (run whatever to trigger these events, and validate they work as expected):

- ./tracee-ebpf -t e=vfs_write,magic_write
- ./tracee-ebpf -t e=socket_dup

## Final Checklist:

Pick "Bug Fix" or "Feature", delete the other and mark appropriate checks.

- [ ] I have made corresponding changes to the documentation.
- [x] My code follows the style guidelines (C and Go) of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented all functions/methods created explaining what they do.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix, or feature, is effective.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published before.

## Git Log Checklist:

My commits logs have:

- [x] Subject starts with "subsystem|file: description".
- [x] Do not end the subject line with a period.
- [x] Limit the subject line to 50 characters.
- [x] Separate subject from body with a blank line.
- [x] Use the imperative mood in the subject line.
- [x] Wrap the body at 72 characters.
- [x] Use the body to explain what and why instead of how.

### CODE REVIEW

File: pkg/ebpf/events_definitions.go [hunk 1]
```diff
@@ -6139,10 +6138,7 @@ var EventsDefinitions = map[int32]EventDefinition{
 		ID32Bit: sys32undefined,
 		Name:    "socket_dup",
 		Probes:  []probe{},
```
Comment (NDStrahilevitz) [inline pkg/ebpf/events_definitions.go:6140]: Shouldn't the probes from the `DupEvent`s be flattened here the same as magic_write?

Comment (yanivagman) [reply pkg/ebpf/events_definitions.go:6140]: No. These are tail calls, not probes.
We currently explicitly install the kprobes in tracee.go.
In a future PR, we should handle tail-calls dependencies of events.

**Comment by rafaeldtinoco** [comment]
> Currently, in order to define which probes are required for an event to
> function correctly, we either define the probes explicitly or use an
> event dependency, which will load and attach the required probes.
> This other method of defining the probes implicitly should be used only
> when the event is a real dependency, and not just as a mechanism to load
> and attach the other event's bpf program.

Should we describe that the "real dependencies" RULE is:

1. events that are derived from event described (eg: CgroupMkdir, CgroupRmdir)
2. events that require processing in processEvents (eg: FinitModuleEvent, InitModuleEvent)

**Review by rafaeldtinoco — APPROVED** [review]
I'm approving but left couple of comments so you can chose to accept if you think they're worth. Overall change looks good and I like the fact that we're auto-loading all programs (we will errors independently of given cmdline).

I see that you are mentioning we should load (or not) programs based on existing kernel features (and chose to attach or not based on current config, that might do reconfigs).

NOTE: Check also tracee.go for network programs "autoload" set to false. Maybe you want to get rid of that as well (so network programs are also always loaded).

File: pkg/ebpf/tracee.go [hunk 2]
```diff
@@ -722,6 +720,9 @@ func (t *Tracee) populateBPFMaps() error {
 	errs = append(errs, t.initTailCall(tailSendBin, "prog_array", "send_bin"))
 	errs = append(errs, t.initTailCall(tailSendBinTP, "prog_array_tp", "send_bin_tp"))
 	errs = append(errs, t.initTailCall(tailKernelWrite, "prog_array", "trace_ret_kernel_write_tail"))
+	errs = append(errs, t.initTailCall(uint32(DupEventID), "sys_exit_tails", "sys_dup_exit_tail"))
+	errs = append(errs, t.initTailCall(uint32(Dup2EventID), "sys_exit_tails", "sys_dup_exit_tail"))
```
[the remark this answers is not in the document]
Comment (yanivagman) [reply pkg/ebpf/tracee.go:724]: Actually, this piece of code only initializes a bpf map ("sys_exit_tails"), and even if the events don't exists for a given architecture, it shouldn't fail

**Comment by yanivagman** [comment]
> Should we describe that the "real dependencies" RULE is:
> 
> 1. events that are derived from event described (eg: CgroupMkdir, CgroupRmdir)
> 2. events that require processing in processEvents (eg: FinitModuleEvent, InitModuleEvent)

Yes, that's correct.

> NOTE: Check also tracee.go for network programs "autoload" set to false. Maybe you want to get rid of that as well (so network programs are also always loaded).

The load of bpf programs happens in initBPF, by calling the function `t.bpfModule.BPFLoadObject()`. There are no other places in the code that load bpf programs (that's not the case for ATTACHING a bpf program)

File: pkg/ebpf/events_definitions.go [hunk 3]
```diff
@@ -23,8 +23,7 @@ type probe struct {
 }
 
 type eventDependency struct {
```
[the remark this answers is not in the document]
Comment (yanivagman) [reply pkg/ebpf/events_definitions.go:25]: There's one more case I would like to consider before eliminating this struct.
I thought about declaring which arguments of an event we depend on are of interest for our event.
This way, we might choose to submit some extra arguments not included in the original event, or only a subset of the other event's arguments.

File: pkg/ebpf/tracee.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (rafaeldtinoco) [reply pkg/ebpf/tracee.go:724]: Yep, you're right. It won't fail here because it is just the tail calls map. Thanks.

**Comment by yanivagman** [comment]
> > NOTE: Check also tracee.go for network programs "autoload" set to false. Maybe you want to get rid of that as well (so network programs are also always loaded).
> 
> The load of bpf programs happens in initBPF, by calling the function `t.bpfModule.BPFLoadObject()`. There are no other places in the code that load bpf programs (that's not the case for ATTACHING a bpf program)

Oops. Just saw that indeed I missed those (not sure why). Let me fix that.

**Comment by yanivagman** [comment]
> NOTE: Check also tracee.go for network programs "autoload" set to false. Maybe you want to get rid of that as well (so network programs are also always loaded).

Ok, so if we do that, CAP_NET_ADMIN will always be required by tracee (to load the tc programs), while if we set autoload to false for these programs, it won't be needed while not required.

**Comment by rafaeldtinoco** [comment]
> Ok, so if we do that, CAP_NET_ADMIN will always be required by tracee (to load the tc programs), while if we set autoload to false for these programs, it won't be needed while not required.

Hum, that's true indeed.

**Pull request merged**